std/stdio.d

```diff
@@ -5457,18 +5455,61 @@ private struct ReadlnAppender
     }
 }
 
-// Private implementation of readln
-private size_t readlnImpl(FILE* fps, ref char[] buf, dchar terminator, File.Orientation orientation)
+private struct LockedFile
 {
-    version (DIGITAL_MARS_STDIO)
+    private @system FILE* fps;
+
+    this(FILE* fps) @trusted
     {
+        this.fps = fps;
         _FLOCK(fps);
-        scope(exit) _FUNLOCK(fps);
+    }
+
+    @disable this();
+    @disable this(this);
+    @disable void opAssign(LockedFile);
 
-        /* Since fps is now locked, we can create an "unshared" version
-         * of fp.
-         */
-        auto fp = cast(_iobuf*) fps;
+    /* Since fps is now locked, we can create an "unshared" version
+     * of fp.
+     */
+    @trusted fp() return scope => cast(_iobuf*) fps;
+
+    ~this() @trusted
+    {
+        _FUNLOCK(fps);
+    }
+}
+
+@safe unittest
+{
+    _iobuf* f() @safe
+    {
+        FILE* fps;
+        auto lf = LockedFile(fps);
+        scope fp = lf.fp;
+        static assert(!__traits(compiles, lf = LockedFile(fps)));
+        static assert(!__traits(compiles, { if (fps) return fp; }));
+        version (ShouldFail)
+        {
+            lf.destroy; // fail in @safe because of @system field?
+            auto x = fp; // demonstrates use after unlock
+            lf.fps = null; // @system field
+        }
+        return null;
+    }
+}
+
+// Private implementation of readln
+private size_t readlnImpl(FILE* fps, ref char[] buf, dchar terminator, File.Orientation orientation) @safe
+{
+    // fp is not shared so these calls are safe
+    alias trusted_FGETC = (scope _iobuf* fp) @trusted => _FGETC(fp);
+    alias trusted_FGETWC = (scope _iobuf* fp) @trusted => _FGETWC(fp);
+
+    version (DIGITAL_MARS_STDIO)
+    {
+        auto lf = LockedFile(fps);
+        auto fp = lf.fp;
```

Comment:
This defeats the purpose of the struct. The struct can't just be a wrapper for the type itself, it needs to be a wrapper for the *functionality that should be trusted*. So like instead of `trusted_FGETWC`, the struct should have a method on it that is trusted.

Reply:
Ok, but I'm not sure how this defeats the purpose of the struct. What is actually unsafe if fp can only be accessed when the file is locked (aside from the destroy issue)?

Reply:
```d
FILE * fp;
{ auto lf = LockedFile(fps); fp = lf.fp; }
```

Reply:
In other words, you should not be able to use the unlocked FILE * if the struct no longer exists.

Reply:
(I plan to move `trusted_FGETC` to a struct method as that's better anyway).

`lf.fp` does not return a `FILE*`, it's a `_iobuf*`. On Windows I get:
```
Error: cannot implicitly convert expression `lf.fp()` of type `_iobuf*` to `shared(_iobuf)*`
```
But all the other code paths use a cast too so I assume they would trigger the same error. Aside from the type error, the assignment `fp = lf.fp` should fail due to -dip1000 but it doesn't:
https://forum.dlang.org/post/cvrzlycycprixbadbpgg@forum.dlang.org

Reply:
OK, I didn't know what `fp` was returning, this wasn't my point though, I should have been more general in the example:
```d
typeof(LockedFile.fp) fp;
{auto lf = LockedFile(fps); fp = lf.fp; }
```
The point is, if you just expose the pointer then it defeats the purpose of encapsulating the trusted behavior.